rfd/9999-webauthn-support.md

```diff
@@ -0,0 +1,501 @@
+CredentialAssertionResponse (modeled as protos).
+
+__MFARegisterChallenge__ and __MFARegisterResponse__, similarly to their
+authenticate counterparts, must change to accomodate a
+[CredentialCreation](https://pkg.go.dev/github.com/duo-labs/webauthn/protocol#CredentialCreation)
+and a
+[CredentialCreationResponse](https://pkg.go.dev/github.com/duo-labs/webauthn/protocol#CredentialCreationResponse)
+(modeled as protos).
+
+All relevant AuthService RPCs (or the methods they rely on) are to be modified
+to support WebAuthn (AddMFADevice, DeleteMFADevice, GenerateUserSingleUseCerts).
+
+#### Backwards compatibility
+
+At the API level (Proxy API, Auth API), backwards compatibility is achieved by
+keeping the old endpoints in place. Additions to existing messages are unknown
+by legacy code, thus ignored by older Proxy and `tsh` versions.
+
+The `/webapi/ping` response may be modified to signalize to `tsh` that the Proxy
+supports the new /mfa/ endpoints (with the assumption, by the Proxy, that the
+Auth server also supports the new endpoints). This allows newer `tsh` versions
+to safely talk to older Proxies (at least for a few releases).
+
+gRPC backwards compatibility is not a concern.
+
+Ping response changes (unrelated content omitted):
+
+```json
+{
+  "auth": {
+    "supports_mfa_endpoints": "yes"
+  }
+}
+```
+
+#### WebAuthn storage
+
+Device storage uses the existing \*MFADevice\* methods in
+[Identity](https://github.com/gravitational/teleport/blob/185e5fda35f3b8fb6debd46acd847cb8250b8f86/lib/services/identity.go#L124),
+no changes required - the system already supports multiple
+[device types](https://github.com/gravitational/teleport/blob/185e5fda35f3b8fb6debd46acd847cb8250b8f86/api/types/types.proto#L1599).
+WebAuthn simply adds a WebAuthnDevice to MFADevice (taking care to notice, in
+documentation, the differences in public key encoding between U2F and WebAuthn.)
+
+WebAuthn requires the persistence of SessionData between challenge attempts
+(for both login and registration). The following additions are proposed:
+
+```go
+type Identity interface {
+	// UpsertWebAuthnSessionData upserts WebAuthn SessionData for the purposes
+	// of verifying a later authentication challenge.
+	// Upserted session data expires according to backend settings.
+	UpsertWebAuthnSessionData(user, sessionID string, sd *SessionData) error
+
+	// GetWebAuthnSessionData retrieves a previously-stored session data by ID,
+	// if it exists and has not expired.
+	GetWebAuthnSessionData(user, sessionID string) (*SessionData, error)
+
+	// DeleteMFAAuthenticateChallenge deletes session data by ID,
+	// if it exists and has not expired.
+	DeleteWebAuthnSessionData(user, sessionID string) error
+}
+```
+
+The key space is `web/users/{user_id}/webauthnsessiondata/{session_id}`.
+Implementation may choose to use hard-coded session IDs for login and
+registration, limiting the possibility of concurrent attempts, or pick distinct
+IDs, as appropriate.
```

Comment:
u2f uses the `user_id` without `session_id` to store the challenge
it seems to work well enough despite limiting concurrent logins

Reply:
Ack.

I'll keep this one - code complexity is pretty much the same, it's not difficult to read and it allows us to fine-tune behavior better.